File: Data Structures & Algorithms/duplicate-integer/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {boolean}
     */
    hasDuplicate(nums) {
        // create a hash
        const hash = new Set();

        // iterate nums
        for (const num of nums) {

        // set the value and compare the value and the hash
          // if you find the duplicate, return true
            if (hash.has(num)) {
                return true;
            }

            hash.add(num);
        }

        
        // The outside of the loop, return false
        return false;
    }
}

// Time complexity: O(n)
// Space complexity: O(n)

File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        lastIndex = 1
        for num in nums:
            res = res + [x.copy() for x in res]
            while lastIndex < len(res):
                res[lastIndex].append(num)
                lastIndex += 1

        return res
# [1, 2]
# [], [1], [2], [1, 2]

# [1, 2, 3]
# [], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]

# 1, 2, 3
# []
# [], [1]
# [], [1], [], [1]